text/0061-functions.md

```diff
@@ -0,0 +1,412 @@
+when {
+  eq(principal,resource.owner) ||
+  eq(principal.org,"Admin")
+};
+```
+
+Likewise, any static analysis tools would work via inlining.
+
+### Templates
+Macros use the same syntax for their parameter variables as do templates, since in both cases variables are "holes" that are filled in with a Cedar expression to make a full construct -- for templates, the construct is a policy, for macros it is an expression.
+
+Templates can include calls to macros. For example:
+
+```
+def semver(?major, ?minor, ?patch) ... // same as initial example
+def semverGT(?lhs, ?rhs) ... // same as initial example
+permit (
+  principal == ?principal,
+  action == Action::"view",
+  resource in ?resource)
+when {
+  semverGT(resource.apiVersion, semver(2,1,0))
+};
+```
+
+When we link this template, we get a full policy, as usual.
+
+A macro parameter can end up having the same name as a template variable, with the effect that it shadows it, which is expected with normal lexical scoping:
+
+```
+def isOwner(?principal,?resource) ?principal == ?resource.owner;
+
+permit(principal,action,resource in ?resource)
+when {
+  isOwner(principal,resource)
+};
+```
+
+Here, notice that the `isOwner` macro's reference to `?resource` is to its parameter, not to the template slot. (We'd recommend to users to choose different parameter names to avoid confusion.)
+
+As already mentioned, you cannot refer to a parameter in a macro that it does not bind, which enforces hygiene. So the following is not allowed.
+
+```
+def isOwner() ?principal == ?resource.owner; 
+// ERROR: cannot refer to unbound variables ?principal, ?resource
+```
+
+
+## Drawbacks
+
+### Redundancy
+Cedar macros can only accomplish things that can already accomplished with Cedar expressions.
+This means we are not expanding the expressive power of Cedar in any way. 
+We are also adding more than one way to accomplish the same task.
+Bringing back our SemVer example:
+```
+// Permit access when the api is greater than 2.1
+permit (principal, action == Action::"someAction", resource)
+when {
+    resource.apiVersion.major > 2 ||
+    (resource.apiVersion.major == 2 && resource.apiVersion.minor >= 1)
+};
+```
+This policy does accomplish the user's goal of encoding the SemVer relationship.
+The problem is readability. A reader of this policy has to work out that it is implementing the standard semantic version comparison operator, and not some bespoke versioning scheme. Another problem is maintainability. If you have multiple policies that reason about SemVers, you have to repeat the logic, inline, in each policy. This violates basic software engineering tenets (Don’t Repeat Yourself), allowing bugs to sneak in via typos or botched copy/paste.
+
+### No custom parsing or error-handling
+Extension functions provide more full-featured constructors through custom parsing and error handling, but Cedar functions provide no such facilities. This may make them harder to read, write, and understand.
```

Comment:
```suggestion
Extension functions provide more full-featured constructors through custom parsing and error handling, but Cedar macros provide no such facilities. This may make them harder to read, write, and understand.
```